drivers/t5838/t5838_nrfx_pdm.c

```diff
@@ -1155,9 +374,12 @@ static int t5838_trigger(const struct device *dev, enum dmic_trigger cmd)
 			drv_data->stopping = true;
 			nrfx_pdm_stop();
 		}
-		if (drv_data->aad_enabled_mode != T5838_AAD_SELECT_NONE) {
-			prv_t5838_enter_sleep_with_AAD(dev);
+#ifdef CONFIG_T5838_AAD_TRIGGER
+
+		if (drv_data->aad_child_dev != NULL) {
+			prv_enter_sleep_with_AAD(drv_data->aad_child_dev);
 		}
+#endif
```

Comment:
I am not sure, if I am satisfied with the fact that `aad_child_dev` is used to dynamically change. I my mind (and how Zephyr uses this) relationships between `struct device`s are compile time fixed, not dynamically changed at runtime. 

However I currently see no any other way of doing this with two separate devices (one for dmic, one for AAD). IMO, it would be better that, if this was one device, but since it is done like that than it can stay.

But please change the `prv_enter_sleep_with_AAD`. To me it looks like that you are creating an access to a private function from another module from thin air. Check how this is done in the `lis2dw12` driver: 
* Declarations of trigger related functions are ifdefed in the header file.
* functions are namespaced - in our case you need `t5838_aad_sleep` (noun->verb to be consistent with the rest)
* You call the function via ifdef, like you already do.

In that case the above forward declaration is redundant and can be removed.

Reply:
Yes, I was struggling to find better solution for this too, but also came to conclusion that this is the best solution possible.

My initial thoughts were also that triggering and PDM should be handled as one device, it was also initially implemented that way. 

I changed to implementation where trigger and PDM are separate with following in mind:
-by separating this into two devices, we can keep original driver in pretty much original/slightly modified state, making it easy to upgrade to further upstream version when original driver gets updates.

-Since there can be situation in the future where single PDM interface has two microphones on the bus implementing this as single device would REALLY complicate things. I know this is not needed now, but I just don't want to risk hitting this wall in the future if that is needed. Doing stuff that way is much more flexible and sensible when we count potential stereo configuration in the future.

-This is the only place where we actually need to access child device from parent and it is only done for convenience sake - it could be also implemented to be manually handled by the user. I think however, this gives us unnecessary point of failure when using driver. What this call does is that after sampling PDM it calls this function that provide microphone with 100kHz clock for some defined time which is needed for correctly setting device into its sleep mode. Not calling this function after sampling results in extra power consumption from the microphone, triggers still work however. Expecting user to manually call this after stopping PDM would probably result in situation where driver user would forget/not know they have to call this every time they stop PDM sampling and then wonder why their device draws more power than expected.


I was considering doing it the way it is done in `lis2dw12` but wasn't sure I should actually expose this function (yes it is ifdef-ed in header file there, but that doesn't really make it private). I am not exactly sure I want to do so, since the way driver is written now that function has no use for the user since it gets called by the driver when it is supposed to. Since this function was ifdef-ed here right above this trigger function I think it is obvious enough where it comes from, so not sure I would call that "from thin air".
I might just expose this function though, since it shouldn't really do any harm.